### PR TITLE
Tested Backend API - Refactor to Client Side Rendering

### PR DESCRIPTION
**Pull request by RonoITU**
From reports on the state of our project so far. Here is a major pull request addressing the following:

### Annotations.API

Solved the following tasks from May 9th report:

- [x] Consider retiring or moving `/images/APITest` endpoint. _Moved to group of testing endpoints._

- [x] `/images/{imageId}` must return 404 if image is not found. 
- [x] `/images/datasets/{datasetId}` must return 404 if dataset not found.
- [x] `/images/upload` must return 201 Created, data about newly created resource. 

- [x] Model for overview of datasets. Should contain only a count of images, matching the front end application. 
	- [x] Omit count of images if not certain to get $O(\log N)$ or better runtime obtaining count. $O(1)$ _solution made._
- [x] `/images/filter/{category}` Make possible to use for lazy loading / pagination. 
	- Strongly consider only returning image metadata. IDs most importantly. 
- [x]  `/images/datasets/{datasetId}` Make possible to use for lazy loading / pagination. 
	- Strongly consider only returning image metadata. IDs most importantly. 
- [x] Create endpoints for getting direct access to images in standard media format. 

- [x] `/images/datasets/{datasetId}` should serve metadata about dataset. 
- [x] Make image datasets fully functional. 
- [x] Make image upload endpoint register additional metadata about upload. 
- [x] Make image upload endpoint create unique IDs correctly for new image resources. 
- [x] Create basic data about users and their use of the service. 
- [x] Change database to be a persistent solution instead of in-memory. 
- [x] Get rid of all "hardcodes" in backend. 

A few potential additions have been left out for now. 

While this effort began on Saturday, the project has been rebased onto main to ensure a linear history. 

### Annotations.API.Tests

- Going from 0 to now 33 unit tests covering services for datasets, images and users. 
- Yet to build tests for image annotations. This will take an additional code review and agreement on functions to be proven. 

### Annotations.Blazor

Refactored to use client side rendering as known from many other web frameworks. (Migration to web-assembly.) This rids us of problems with technical debt related to the use of Blazor Server Interactive. 

This action moves almost all of our web pages into a new client project. 

Contact with the API is accomplished using the Backend For Frontend pattern. The server is configured to act as a proxy for all calls to the backend, and is responsible for attaching the necessary token for authentication. This has been hardcoded to localhost for now. It is trivial to add a user secret or otherwise allow the backend API domain to be set at runtime. 

Client side services have been written where relevant for retrieving object models from the API. Other types of data, namely multimedia, can be requested directly by the HTML element with no additional code except for the proxy redirect already in place. 

### Annotations.Blazor.Tests

- I have ensured that all existing B-Unit tests work with this project. 
- For the E2E Playwright tests, there should not be problems, but I cannot guarantee. 

### Annotations.Core

- As suggested previously, I have added basic user data functionality by first simplifying the `User` entity and model. This has been tested through and through, integrated with the rest of the entities and models where required. 
- Datasets are now a functional entity with essential API support. 
- The image entity and model is now solely for metadata. Azure storage handles the media files and their type directly. The link between these two services is maintained by the image id. 
- Images and datasets have "soft delete" implemented. First and foremost this ensures that a new entry cannot replace a previously deleted one by accident. 
- There are minor choices and explanations I will have to provide for them. 

## New details on how to run

- New user secret. An SQLite database will be created/used based on the connection string provided. I.e.: 
  - `ConnectionStrings:DefaultConnection = DataSource=C:\temp\annotations.db`
- You will need to use this command to create a local test token with correct user information: 
  - `dotnet user-jwts create --name TestUser --claim "Sub=123abcdefg456"`
- Error messages from the code running on the client are available using the developer tools in browser. Hit `F12` to access.

### CODE REVIEW

**Comment by RonoITU** [comment]
_Summary using GitHub Copilot AI_

The changes to the API project can be summarized as follows:

### 1. **Test Coverage Enhancements**
- **New Test Classes Added**:
  - `DatasetBuilderTests`: Tests for creating datasets with valid parameters and handling invalid cases (e.g., missing user, title, or category).
  - `DatasetServiceTests`: Tests for key `DatasetService` methods such as retrieving datasets, setting images, and handling edge cases like missing datasets or images.
  - `ImageServiceTests`: Tests for image retrieval and deletion, including edge cases like non-existent images.
  - `ImageUploaderTests`: Tests for uploading images, including validations like missing categories.
  - `UserServiceTests`: Tests for user-related functionality, such as retrieving all users, finding users, and creating users.

- **Improved Documentation**:
  - Added XML documentation (`/// <summary>`) for all test methods, describing their purpose and behavior.
  - Example: Tests like `OnNoUser_ThrowsException` and `OnNoTitle_BadRequest` now have clear descriptions.

### 2. **Refactoring and Reorganization**
- **Namespace Changes**:
  - The `Annotations.API.Services` namespace was reorganized:
    - `AnnotationService` and related utilities moved to `Annotations.API.Services.ImageAnnotation`.
    - Image-related services moved to `Annotations.API.Services.Images`.
    - Dataset and user services moved to `Annotations.API.Services.Datasets` and `Annotations.API.Services.Users`, respectively.

- **Endpoint Refactoring**:
  - Introduced new endpoint classes (`DatasetEndpoints`, `ImageEndpoints`, `UserEndpoints`, `TestingEndpoints`) under `Annotations.API.Endpoints`.
  - Removed legacy endpoint files like `Groups/ImageEndpoints.cs` and `Groups/UserEndpoints.cs`.

### 3. **Database Changes**
- **Entity Framework Updates**:
  - Replaced `IdentityDbContext` with a custom `DbContext` for `AnnotationsDbContext`.
  - Removed ASP.NET Identity-related tables (`AspNetUsers`, `AspNetRoles`, etc.).
  - Simplified user management with a custom `Users` table.
  - Added `DatasetEntry` for linking datasets and images with additional metadata like `OrderNumber`.

- **Migration Updates**:
  - Introduced new migrations (`20250513135858_ImageCount`) to reflect schema changes like the removal of Identity tables, addition of `DatasetEntry`, and refactoring dataset and image entities.

### 4. **Service Updates**
- **Dataset Service**:
  - Added functionality for setting images in a dataset with validations and improved error handling.
  - Introduced `ModifyDatasetResult` for standardizing responses when modifying datasets.

- **Image Service**:
  - Improved image upload and retrieval logic.
  - Added support for handling deleted images gracefully.

- **User Service**:
  - Simplified user management by working directly with JWT claims and the custom `Users` table.

### 5. **API Improvements**
- **Routing Changes**:
  - New routes introduced for `/datasets`, `/testing`, `/images`, and `/users`, with methods mapped in endpoint classes.

- **Configuration Updates**:
  - Updated `Program.cs` to reflect changes in service registrations and database migrations.
  - Removed the temporary SQLite database setup and initialized a proper database connection using a connection string.

### 6. **Testing Enhancements**
- **Testing Endpoints**:
  - Added `TestingEndpoints` for testing API functionality, including a sample `/tryme` endpoint.

- **Comprehensive Error Coverage**:
  - Tests now cover scenarios like missing datasets, invalid requests, and edge cases for dataset and image operations.

### Summary
The project has significantly improved test coverage and modularity. The new tests ensure robust validation of datasets, images, users, and annotations while handling edge cases. The refactoring enhances maintainability through organized namespaces and endpoint classes. The database changes simplify user and dataset management, making the API more focused and efficient.

**Comment by RonoITU** [comment]
_Summary of the frontend changes made using GitHub Copilot AI_

Here is a summarized overview of the changes made to the Blazor frontend project reflecting the transition from Server-Side Interactive Mode to Client-Side WebAssembly:

---

### **Key Changes**
1. **Transition to WebAssembly Client-Side Rendering**
   - The project has moved from **InteractiveServer** mode (SignalR-based server-side rendering) to **InteractiveWebAssembly** mode.
   - This transition is evident in updates to files like `App.razor`, `Program.cs`, and usage of `InteractiveWebAssemblyRenderMode` instead of `InteractiveServerRenderMode`.
   - Removed `@rendermode InteractiveServer` directives from `.razor` files.

2. **Backend-for-Frontend (BFF) Pattern Implementation**
   - The frontend server now acts as a **proxy** for API requests, authenticating users with cookies while attaching JWT tokens to backend API requests.
   - Introduced a new endpoint `/api/{**remainder}` in `Program.cs`, which uses **YARP (Yet Another Reverse Proxy)** for forwarding requests to the backend API with user JWT tokens attached.
   - Added middleware for token-based request transformation within the proxy.

3. **Authentication and Authorization Updates**
   - Introduced `PersistingAuthenticationStateProvider` for maintaining authentication state in WebAssembly mode.
   - Updated the authentication mechanism to use cookies for the frontend and JWT tokens for backend communication.
   - Refactored authorization attributes in `.razor` files to restrict access based on roles (`[Authorize(Roles = "...")]`).
   - Added `Microsoft.AspNetCore.Components.Authorization` and related settings to the program.

4. **Dependency Injection and Services Refactoring**
   - Replaced usage of `IAPIServices` with `IAnnotationDataAccessor` across components for handling API interactions.
   - Services now instantiate `AnnotationDataAccessor` and other WebAssembly-compatible services.
   - Removed the `APIServices` implementation altogether.

5. **Project Structure Reorganization**
   - Created a dedicated `Annotations.Blazor.Client` project, splitting UI components and services into this separate client-side project.
   - Moved various Razor components and services (e.g., `AnnotationDataAccessor`, `DatasetAccessor`, etc.) to the new project.

6. **Authorization Enhancements**
   - Added `@using Microsoft.AspNetCore.Authorization` and role-based `[Authorize]` attributes to Razor pages for enhanced access control.
   - Several `.razor` files now restrict access to roles like `Administrator`, `Manager`, and `AnnotationsUser`.

7. **Dependency and Package Updates**
   - Introduced new dependencies in `Annotations.Blazor.csproj` for WebAssembly support and the YARP proxy:
     - `Microsoft.Extensions.ServiceDiscovery.Yarp`
     - `Microsoft.AspNetCore.Components.WebAssembly.Server`
   - Updated existing package references for compatibility with Blazor WebAssembly.

8. **Refactoring Tests**
   - Adjusted unit and E2E tests to reflect changes in services and components:
     - Replaced `IAPIServices` with `IAnnotationDataAccessor` in test setups.
     - Updated namespaces and references for the new client project.

9. **Routing and Layout Adjustments**
    - Refactored routing in `Routes.razor`, adding comments and references to Blazor sample projects.
    - Updated layouts and navigation components to align with the WebAssembly-based architecture.

---

### **Benefits of These Changes**
- **Scalability**: WebAssembly removes the need for SignalR connections, reducing server load.
- **Improved Authentication**: The Backend-for-Frontend pattern ensures secure, token-based distributed system interactions while maintaining cookie-based authentication for users.
- **Modularity**: The split into `Annotations.Blazor.Client` enhances maintainability and separation of concerns.
- **Modernized Architecture**: The project is more aligned with modern Blazor WebAssembly practices, ensuring better performance and compatibility for client-side applications.

This transformation represents a significant modernization of the project, shifting from server-side Blazor to a more scalable and secure WebAssembly architecture using a Backend-for-Frontend pattern.

**Review by Spon-art — COMMENTED** [review]
Note that your Blazor to API is hardcoded to https://localhost:7250/. Consider using environment varibles or another way to route the URI.

Changed the role of testuser@faketestsite.dk as the new setup with Blazor webassembly would trigger a loop when no claims were present for roles. Added dummy/useless guest role in OrchardCore. So E2E test now works again!

Also two nullable warnings. Consider removal or suppression of these.

The new style of having lots of spaces between classes are fine, just be aware if the rest of team agrees on the layout

But the API has really undergone huge changes. I have testet on my machine locally. GJ

image

image

File: Annotations.API.Tests/DatasetServiceTests.cs [hunk 1]
```diff
@@ -0,0 +1,747 @@
...
+        // Assert
+        Assert.Null(result);
+    }
+
+
+
+
+
+    /// <summary>
+    /// Tests the CreateDatasetAsync method to ensure it creates a new dataset.
```
Comment (tomat-suppe) [inline Annotations.API.Tests/DatasetServiceTests.cs:163]: Summary here should be "Tests the DeleteDatasetAsync method to ensure it deletes a dataset".

File: Annotations.API/Endpoints/DatasetEndpoints.cs [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+
+
+
+    /// <summary>
+    /// Finds and returns all images within a certain category.
+    /// </summary>
+    /// <param name="category"></param>
+    /// <param name="imageService"> An image service instance. </param>
+    /// <returns> An array of images as JSON string with the wanted category. </returns>
+    private static async Task<ICollection<ImageModel>> FilterImagesHandler(
```
Comment (tomat-suppe) [inline Annotations.API/Endpoints/DatasetEndpoints.cs:50]: Should we move this method to ImageEndpoints? I know it contains several images, but the filter category doesn't technicaly have any ties to datasets.

File: Annotations.API/Program.cs [hunk 3]
```diff
@@ -98,18 +103,33 @@
...
+app.MapGet("/error", () => "Exception raised in API");
+
 
 // Development/Debugging middleware.
 if (app.Environment.IsDevelopment())
 {
     app.UseSwagger();
     app.UseSwaggerUI();
-    InitializeTempDatabase();
+    //InitializeTempDatabase();
```
Comment (tomat-suppe) [inline Annotations.API/Program.cs:132]: Should this be removed completely? I cannot see the method for it anymore, and it should be unneeded, if I understand the new code correctly

**Review by jakso535 — DISMISSED** [review]
I believe this can be pushed to main. I have some comments that needs to be adressed in the next iteration:

- Datasets deleted through the API still appear on the webpage. They do get marked as deleted as intended, frontend support to show deletion should be added.

- Images deleted through API do get removed in Blazor and get replaced with a standin ? with its image id displayed. Maybe add something else to show instead.

Underneath i have some general notes for next iteration: 
- If a vessel tree has been marked as not visible, the viibiity toggle for points and segments do not work. Consider either adding this functionality, or remove it entirely, so the user only can mark whole trees.

- Lets give the home page some more love, so it can be used as a navigation tool to other parts of the user experience, possibly also adding the user data we now have access to.

File: Annotations.API/Services/Datasets/DatasetService.cs [hunk 4]
```diff
@@ -0,0 +1,306 @@
...
+/// Defines a scoped service for dataset information stored in the backend API database.
+/// </summary>
+public interface IDatasetService
+{
+    /// <summary>
+    /// Retrieves all images within a certain category.
+    /// </summary>
+    /// <param name="category"></param>
+    /// <returns></returns>
+    Task<ICollection<ImageModel>> GetFilteredImageSetAsync(string category);
```
Comment (tomat-suppe) [inline Annotations.API/Services/Datasets/DatasetService.cs:20]: Copy comment from the DatasetEndpoint class: 
"Should we move this method to ImageEndpoints? I know it contains several images, but the filter category doesn't technicaly have any ties to datasets."

File: Annotations.API/Services/Datasets/DatasetService.cs [hunk 5]
```diff
@@ -0,0 +1,306 @@
...
+
+        return HttpStatusCode.NoContent;
+    }
+
+
+
+
+
+    /// <summary>
+    /// Sets the images for a dataset. This will replace all existing images in the dataset with the provided image IDs.
```
Comment (tomat-suppe) [inline Annotations.API/Services/Datasets/DatasetService.cs:160]: Is there a reason for replacing all images instead of appending upon the datasets existing images+imagecount?

File: Annotations.API/Services/Datasets/DatasetService.cs [hunk 6]
```diff
@@ -0,0 +1,306 @@
...
+        {
+            UserName = user.UserName,
+        };
+    }
+
+
+
+
+
+    private static DatasetModel ToDatasetModel(Dataset dataset)
```
Comment (tomat-suppe) [inline Annotations.API/Services/Datasets/DatasetService.cs:288]: Why don't we use the Annotations.Core.Models DatasetModel here?

File: Annotations.API/Services/Datasets/ModifyDatasetResult.cs
Comment (tomat-suppe) [inline Annotations.API/Services/Datasets/ModifyDatasetResult.cs:1]: In Services/Images/Imageservice (and Annotations.API/Services/Images/ImageUploader.cs) the corresponding to this class, sealed class GetImageResult is with the rest of the imageservice class, but here it is divorced from the DatasetService. I have no strong feelings either way, but maybe we want to make it the same

File: Annotations.API/Services/Images/ImageUploader.cs
Comment (tomat-suppe) [inline Annotations.API/Services/Images/ImageUploader.cs:1]: Last 4 methods are un-documented, can we get documentation on these?

File: Annotations.Blazor.Client/Services/AnnotationDataAccessor.cs
Comment (tomat-suppe) [inline Annotations.Blazor.Client/Services/AnnotationDataAccessor.cs:1]: No documentation in this class.

File: Annotations.Core/Entities/EntitiesDiagram.cd
Comment (tomat-suppe) [inline Annotations.Core/Entities/EntitiesDiagram.cd:1]: What is this? A diagram for a required plug-in or something else? :)

**Review by tomat-suppe — CHANGES_REQUESTED** [review]
Comments on things I would wish for changes on are attached to this review.


There is one or two comments about un-documented classes. Because of the set-up of the changes in this branch, I cannot see if these are new or old classes, and therefore I have added the comments regardless.

The 2-ish classes that are very authentication specific are convoluted for me to understand - I do not feel I can give proper review on those :)

**Comment by RonoITU** [comment]
Thank you all for the reviews today. There were some interesting points that I will look into ASAP. Expect to see additional pushes before tomorrow.

File: Annotations.API.Tests/DatasetServiceTests.cs [hunk 7]
```diff
@@ -0,0 +1,747 @@
...
+        using (var context = new AnnotationsDbContext(options))
+        {
+            var datasetService = new DatasetService(context);
+            result = await datasetService.SetImagesAsync(1, [1, 2, 3]);
+        }
+
+
+        // Assert
+        Assert.Equal(201, result.StatusCode);
+        Assert.Equal(3, result.Dataset.ImageIds.Count);
```
[the remark this answers is not in the document]
Comment (RonoITU) [reply Annotations.API.Tests/DatasetServiceTests.cs:544]: Here is a deliberate stray away from perfect database normalization. For some applications, i.e. the image overview, I want to have a count available in constant time, without running an expensive "include" statement in the query.

File: Annotations.API.Tests/ImageUploaderTests.cs [hunk 8]
```diff
@@ -0,0 +1,381 @@
...
+        Assert.NotEqual(string.Empty, uploadResult.Error);
+    }
+
+
+
+
+
+    /// <summary>
+    /// Test to see that it is not possible to upload an image without a user.
+    /// <br />
```
[the remark this answers is not in the document]
Comment (RonoITU) [reply Annotations.API.Tests/ImageUploaderTests.cs:229]: I will resolve this by replacing with a simpler statement.

File: Annotations.API/Endpoints/DatasetEndpoints.cs [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+    /// <param name="pathBuilder"></param>
+    public static void MapEndpoints(RouteGroupBuilder pathBuilder)
+    {
+        // Anti-forgery is disabled. This was decided because the backend will not serve any forms.
+        // Anti-forgery measures are covered in the front-end, and by the JWT token protection. 
+        pathBuilder.RequireAuthorization().DisableAntiforgery();
+
+        pathBuilder.MapGet("/filter/{category}", FilterImagesHandler);
+
+        pathBuilder.MapGet("/overview", RetrieveAllDatasetHandler);
```
[the remark this answers is not in the document]
Comment (RonoITU) [reply Annotations.API/Endpoints/DatasetEndpoints.cs:28]: I will resolve this, and make the change.

File: Annotations.API/Endpoints/DatasetEndpoints.cs (hunk 2 above)
Comment (RonoITU) [reply Annotations.API/Endpoints/DatasetEndpoints.cs:50]: The reason for this was that the application had similarities with datasets. I can resolve this by moving the method in both endpoint system and service. I will do this and add an additional endpoint to retrieve the metadata for a single image.

File: Annotations.API/Program.cs (hunk 3 above)
Comment (RonoITU) [reply Annotations.API/Program.cs:132]: This is an artifact that I forgot about. I will remove this to resolve.

File: Annotations.API/Services/Datasets/DatasetService.cs (hunk 4 above)
Comment (RonoITU) [reply Annotations.API/Services/Datasets/DatasetService.cs:20]: I will address this and comment when resolved.

File: Annotations.API/Services/Datasets/ModifyDatasetResult.cs
Comment (RonoITU) [reply Annotations.API/Services/Datasets/ModifyDatasetResult.cs:1]: The reason is that `DatasetBuilder` and `DatasetService` had similar needs, and were able to share the `ModifyDatasetResult` class. 

The needs of the `ImageUploader` and the `ImageService` were different. One needs to return limited metadata, the other needs to return the full image for downloading. 

We could argue whether the image uploader should return the full `ImageModel` metadata as needed.

File: Annotations.API/Services/Images/ImageUploader.cs
Comment (RonoITU) [reply Annotations.API/Services/Images/ImageUploader.cs:1]: I will add this to the private methods as requested to resolve.

File: Annotations.Core/Entities/EntitiesDiagram.cd
Comment (RonoITU) [reply Annotations.Core/Entities/EntitiesDiagram.cd:1]: Yes. Visual Studio has a tool to create class diagrams. This is a diagram of the entities. Easiest to maintain and share using the codebase.

File: Annotations.Blazor.Client/Services/AnnotationDataAccessor.cs
Comment (RonoITU) [reply Annotations.Blazor.Client/Services/AnnotationDataAccessor.cs:1]: I will resolve this generally for the client service classes by adding documentation to the interfaces.

**Comment by RonoITU** [comment]
These new commits address the following: 

- [x] _Datasets deleted through the API still appear on the webpage. They do get marked as deleted as intended, frontend support to show deletion should be added._ Resolve by applying a filter in the backend. Use TDD. 
- [x] Move the filtered images endpoint to the image service and image endpoint group. 
- [x] Create an additional endpoint for access to the metadata of a single image. 
- [x] Resolve multiple complaints about lacking documentation in the set of API tests. 
- [x] Resolve multiple complaints about lacking documentation in the front end.

File: Annotations.API/Services/Datasets/DatasetService.cs (hunk 4 above)
Comment (RonoITU) [reply Annotations.API/Services/Datasets/DatasetService.cs]: I have resolved this by moving the endpoint and service.

File: Annotations.API/Endpoints/DatasetEndpoints.cs (hunk 2 above)
Comment (RonoITU) [reply Annotations.API/Endpoints/DatasetEndpoints.cs]: I have resolved this by moving the endpoint and service.

**Comment by RonoITU** [comment]
Just now, I have solved the problem of adjusting rows of images correctly within our image gallery. Using inspect element, I have made sure that loading is indeed lazy to an appropriate extent. 

With this fixed, I am ready to request that you review again.

**Review by Spon-art — APPROVED** [review]
Approved (under the condition of a pass from Sophie and the comments made).
Love the DatasetServiceTests that test that something is no longer present.

**Review by tomat-suppe — APPROVED** [review]
Thank you for the changes, and good work.

The lazy loading is _beautiful_ now also :)
